pkg/controller/zerocapacity/zero_capacity.go

```diff
@@ -0,0 +1,452 @@
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("Unable to fetch %s CR '%s': %w", resource, request.Name, err)
+	}
+
+	phase := instance.Phase()
+	switch phase {
+	case "":
+		return z.initializeResources(request, instance)
+	case ZeroSucceeded, ZeroFailed:
+		return z.cleanupResources(request)
+	}
+
+	infinispan := &v1.Infinispan{}
+	clusterName := instance.Spec().Cluster
+	clusterObjKey := client.ObjectKey{
+		Namespace: namespace,
+		Name:      clusterName,
+	}
+	if err := z.Get(context.Background(), clusterObjKey, infinispan); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("CR '%s' not found: %w", clusterName, err)
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("Unable to fetch CR '%s': %w", clusterName, err)
+	}
+
+	user := consts.DefaultOperatorUser
+	pass, err := users.PasswordFromSecret(user, infinispan.GetSecretName(), namespace, z.Kube)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	httpConfig := ispnclient.HttpConfig{
+		Username:  user,
+		Password:  pass,
+		Namespace: namespace,
+		Protocol:  strings.ToLower(string(infinispan.GetEndpointScheme())),
+	}
+	httpClient := curl.New(httpConfig, z.Kube)
+
+	if phase == ZeroPending {
+		return z.execute(httpClient, request, instance)
+	}
+	// Phase must be ZeroRunning, so wait for execution to complete
+	return z.waitForExecutionToComplete(httpClient, request, instance)
+}
+
+func (z *Controller) initializeResources(request reconcile.Request, instance Resource) (reconcile.Result, error) {
+	ctx := context.Background()
+	name := request.Name
+	namespace := request.Namespace
+	spec := instance.Spec()
+	statefulset := &appsv1.StatefulSet{}
+	clusterKey := k8client.ObjectKey{
+		Namespace: namespace,
+		Name:      spec.Cluster,
+	}
+	if err := z.Client.Get(ctx, clusterKey, statefulset); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{RequeueAfter: consts.DefaultWaitOnCluster}, fmt.Errorf("Unable to load Infinispan Cluster '%s': %w", spec.Cluster, err)
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// If a VolumeSource has not been configured, then we create a new pvc that will not be deleted with the CR
+	if spec.Volume.InitPvc && spec.Volume.VolumeSource == (corev1.VolumeSource{}) {
+		err := z.Client.Create(ctx, &corev1.PersistentVolumeClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+			Spec: corev1.PersistentVolumeClaimSpec{
+				AccessModes: []corev1.PersistentVolumeAccessMode{
+					corev1.ReadWriteOnce,
+				},
+				Resources: corev1.ResourceRequirements{
+					Requests: corev1.ResourceList{
+						// TODO make configurable?
+						corev1.ResourceStorage: consts.DefaultPVSize,
```

Comment:
Shouldn't be this value computed? Even though that probably depends what are we backing up here. Only cache, counters etc. definitions or are cache data also included?

Reply:
Yes I think we should compute the value eventually. A backup can contain any or all of the following: caches+content, cache-configs,counters,scripts,proto-schemas. So for this to be automatic, I will need to add a way to return a size estimate of a backup from the server.

For the initial implementation, I think something basic like setting volume size to `number of pods*configured ram` is suffciient. If a user requires more than this, i.e. if they have a persistent store, it's possible for them to configure a `VolumeSource` with sufficient capacity.